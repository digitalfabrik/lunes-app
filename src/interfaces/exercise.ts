export interface IVocabularyOverviewScreen {
  navigation: any
  route: any
}

export interface IVocabularyTrainerScreen {
  navigation: any
  route: any
}

export interface IAlternativeWordProps {
  alt_word: string
  article: string
}

export interface IDocumentProps {
  id: number
  word: string
  article: string
  document_image: Array<{ id: number; image: string }>
  audio: string
  alternatives: IAlternativeWordProps[]
}

export interface IVocabularyOverviewListItemProps {
  id: number
  word: string
  article: string
  image: string
  audio: string
}

export interface IConfirmationModalProps {
  visible: boolean
  setIsModalVisible: Function
  navigation: any
  extraParams: any
}

export interface IAnswerSectionProps {
  tryLater: () => void
  currentDocumentNumber: number
  setCurrentDocumentNumber: Function
  documents: IDocumentProps[]
  finishExercise: Function
  trainingSet: string
  disciplineTitle: string
}

export interface IFeedbackProps {
  secondAttempt: boolean
  result: string
  document?: IDocumentProps
  input: string
}

export interface IActionsProps {
  tryLater: () => void
  giveUp: () => void
  result: string
  checkEntry: () => void
  getNextWord: () => void
  input: string
  isFinished: boolean
  secondAttempt: boolean
}
