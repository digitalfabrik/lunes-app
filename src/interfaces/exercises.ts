export interface IExercisesScreenProps {
  route: any
  navigation: any
}
