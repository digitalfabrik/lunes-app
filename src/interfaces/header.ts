export interface IHeaderProps {
  top: number | undefined
}
