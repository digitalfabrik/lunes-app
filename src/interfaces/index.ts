import { ReactElement } from 'react'

export interface ILoadingProps {
  children: ReactElement
  isLoading: boolean
}

export interface IButtonProps {
  onPress: () => void
  disabled?: boolean
  children: ReactElement
  theme?: string
  testID?: string
}
export interface IPopoverProps {
  setIsPopoverVisible: (param: boolean) => void
  isVisible: boolean
  children: ReactElement
}
