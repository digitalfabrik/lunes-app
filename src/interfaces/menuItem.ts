import { ReactElement } from 'react'

export interface IMenuItemProps {
  selected: boolean
  onPress: () => void
  icon: string
  title: string
  children: ReactElement
}
