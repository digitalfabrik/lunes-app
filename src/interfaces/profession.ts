export interface IProfessionScreenProps {
  navigation: any
}

export interface IProfessionsProps {
  id: number
  title: string
  description: string
  icon?: string
  total_training_sets: number
}
