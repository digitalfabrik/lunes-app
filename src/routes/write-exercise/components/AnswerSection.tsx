import React, { useState, ReactElement } from 'react'
import { TouchableOpacity, Pressable, Keyboard } from 'react-native'
import { heightPercentageToDP as hp } from 'react-native-responsive-screen'
import stringSimilarity from 'string-similarity'
import styled from 'styled-components/native'

import { CloseIcon } from '../../../../assets/images'
import { ExerciseKeys, SimpleResultType } from '../../../constants/data'
import { DocumentType } from '../../../constants/endpoints'
import labels from '../../../constants/labels.json'
import { COLORS } from '../../../constants/theme/colors'
import AsyncStorage from '../../../services/AsyncStorage'
import Actions from './Actions'
import ArticleMissingPopoverContent from './ArticleMissingPopoverContent'
import Feedback from './FeedbackSection'
import Popover from './Popover'

const StyledContainer = styled.View`
  padding-top: 20px;
  padding-bottom: 30px;
  align-items: center;
  position: relative;
  width: 100%;
  height: 85%;
`
const TextInputContainer = styled.View<{ styledBorderColor: string }>`
  width: 80%;
  height: ${hp('8%')}px;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  border-width: 1px;
  border-radius: 2px;
  padding-right: 15px;
  padding-left: 15px;
  margin-bottom: 5%;
  border-color: ${prop => prop.styledBorderColor};
`
const StyledTextInput = styled.TextInput`
  flex: 1;
  font-size: ${props => props.theme.fonts.largeFontSize};
  font-weight: ${props => props.theme.fonts.lightFontWeight};
  letter-spacing: ${props => props.theme.fonts.listTitleLetterSpacing};
  font-family: ${props => props.theme.fonts.contentFontRegular};
  color: ${prop => prop.theme.colors.lunesBlack};
  width: 90%;
`

export interface AnswerSectionPropsType {
  tryLater: () => void
  currentDocumentNumber: number
  setCurrentDocumentNumber: Function
  documents: DocumentType[]
  finishExercise: Function
  trainingSet: string
  disciplineTitle: string
  setHintsEnabled: (hintsEnabled: boolean) => void
}

const almostCorrectThreshold = 0.6

const AnswerSection = ({
  currentDocumentNumber,
  setCurrentDocumentNumber,
  finishExercise,
  tryLater,
  trainingSet,
  disciplineTitle,
  documents,
  setHintsEnabled
}: AnswerSectionPropsType): ReactElement => {
  const [isArticleMissing, setIsArticleMissing] = useState<boolean>(false)
  // The currently entered answer
  const [input, setInput] = useState<string>('')
  // The last submitted answer
  const [submission, setSubmission] = useState<string | null>(null)
  // The result of the submitted answer
  const [result, setResult] = useState<SimpleResultType | null>(null)
  const [isFocused, setIsFocused] = useState<boolean>(false)

  const touchable: any = React.createRef()

  const document = documents[currentDocumentNumber]
  const totalNumbers = documents.length
  const secondAttempt = !!submission

  const capitalizeFirstLetter = (string: string): string => {
    return string.charAt(0).toUpperCase() + string.slice(1)
  }

  const checkEntry = async (): Promise<void> => {
    const splitInput = input.trim().split(' ')

    if (splitInput.length < 2) {
      setIsArticleMissing(true)
      return
    }

    setSubmission(input)
    setIsArticleMissing(false)

    const article = capitalizeFirstLetter(splitInput[0])
    const word = splitInput[1]

    const newResult = validateAnswer(article, word)
    setResult(newResult)

    if (!secondAttempt) {
      await storeResult(newResult)
    }
    if (newResult !== 'similar') {
      setHintsEnabled(true)
    }
  }

  const validateAnswer = (article: string, word: string): SimpleResultType => {
    const validAnswers = [{ article: document.article, word: document.word }, ...document.alternatives]
    if (validAnswers.some(answer => answer.word === word && answer.article.value === article)) {
      // Article and word are an exact match with either the document or an alternative
      return 'correct'
    } else if (validAnswers.some(answer => answer.word === word)) {
      // Word is an exact match with either the document or an alternative -> just the article is wrong
      return 'similar'
    } else if (
      validAnswers.some(answer => stringSimilarity.compareTwoStrings(answer.word, word) > almostCorrectThreshold)
    ) {
      // TODO Is the article relevant here? Before article was checked for alternatives but not for default word
      // Word is similar to either the document or an alternative
      return 'similar'
    }

    return 'incorrect'
  }

  const giveUp = async (): Promise<void> => {
    const previousResult = result
    setResult('incorrect')
    await storeResult(previousResult ?? 'incorrect')
    setHintsEnabled(true)
  }

  const continueExercise = (): void => {
    setResult(null)
    setSubmission(null)
    setInput('')
    setHintsEnabled(false)

    if (currentDocumentNumber === totalNumbers - 1) {
      finishExercise()
    } else {
      setCurrentDocumentNumber(currentDocumentNumber + 1)
    }
  }

  const storeResult = async (score: SimpleResultType): Promise<void> => {
    try {
      const exercise = (await AsyncStorage.getExercise(ExerciseKeys.writeExercise)) ?? {}
      exercise[disciplineTitle] = exercise[disciplineTitle] ?? {}
      exercise[disciplineTitle][trainingSet] = exercise[disciplineTitle][trainingSet] ?? {}

      exercise[disciplineTitle][trainingSet][document.word] = {
        ...document,
        result: score
      }

      await AsyncStorage.setExercise(ExerciseKeys.writeExercise, exercise)

      const session = await AsyncStorage.getSession()
      if (session === null) {
        throw new Error('Session is not saved correctly!')
      }
      const newSession = {
        ...session,
        retryData: {
          data: documents.slice(currentDocumentNumber + 1, totalNumbers)
        }
      }
      await AsyncStorage.setSession(newSession)
    } catch (e) {
      console.error(e)
    }
  }

  const getBorderColor = (): string => {
    if (isFocused) {
      return COLORS.lunesBlack
    } else if (!secondAttempt && !input) {
      return COLORS.lunesGreyMedium
    } else if (!result && !secondAttempt) {
      return COLORS.lunesBlack
    } else if (result === 'correct') {
      return COLORS.lunesFunctionalCorrectDark
    } else if (result === 'incorrect' || !secondAttempt) {
      return COLORS.lunesFunctionalIncorrectDark
    } else {
      return COLORS.lunesFunctionalAlmostCorrectDark
    }
  }

  const editable = result === null || result === 'similar'

  return (
    <Pressable onPress={Keyboard.dismiss}>
      <StyledContainer>
        <Popover isVisible={isArticleMissing} setIsPopoverVisible={setIsArticleMissing} ref={touchable}>
          <ArticleMissingPopoverContent />
        </Popover>

        <TextInputContainer testID='input-field' ref={touchable} styledBorderColor={getBorderColor()}>
          <StyledTextInput
            placeholder={secondAttempt ? labels.exercises.write.newTry : labels.exercises.write.insertAnswer}
            placeholderTextColor={COLORS.lunesBlackLight}
            value={input}
            onChangeText={setInput}
            editable={editable}
            onFocus={() => setIsFocused(true)}
            onBlur={() => setIsFocused(false)}
            onSubmitEditing={checkEntry}
          />
          {editable && input !== '' && (
            <TouchableOpacity onPress={() => setInput('')}>
              <CloseIcon />
            </TouchableOpacity>
          )}
        </TextInputContainer>

        {result && <Feedback result={result} document={document} submission={submission} />}

        <Actions
          tryLater={tryLater}
          giveUp={giveUp}
          input={input}
          result={result}
          checkEntry={checkEntry}
          continueExercise={continueExercise}
          secondAttempt={secondAttempt}
          isFinished={currentDocumentNumber === totalNumbers - 1}
        />
      </StyledContainer>
    </Pressable>
  )
}

export default AnswerSection
