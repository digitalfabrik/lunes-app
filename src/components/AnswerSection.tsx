import React, { useState } from 'react'
import { View, TouchableOpacity, TextInput, Platform, StyleSheet } from 'react-native'
import { CloseIcon, VolumeUp } from '../../assets/images'
import { COLORS } from '../constants/colors'
import { IAnswerSectionProps } from '../interfaces/exercise'
import Popover from './Popover'
import SoundPlayer from 'react-native-sound-player'
import Tts from 'react-native-tts'
import AsyncStorage from '@react-native-async-storage/async-storage'
import Feedback from './FeedbackSection'
import stringSimilarity from 'string-similarity'
import Actions from './Actions'
import PopoverContent from './PopoverContent'
import { widthPercentageToDP as wp, heightPercentageToDP as hp } from 'react-native-responsive-screen'

export const styles = StyleSheet.create({
  container: {
    paddingVertical: 40,
    alignItems: 'center',
    position: 'relative',
    width: '100%',
    height: hp('65%')
  },
  textInputContainer: {
    width: wp('80%'),
    height: hp('8%'),
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    borderWidth: 1,
    borderRadius: 2,
    paddingHorizontal: 15,
    marginBottom: hp('6%')
  },
  textInput: {
    fontSize: wp('4.5%'),
    fontWeight: 'normal',
    letterSpacing: 0.11,
    fontFamily: 'SourceSansPro-Regular',
    color: COLORS.lunesBlack,
    width: wp('60%')
  },
  volumeIcon: {
    position: 'absolute',
    top: -20,
    left: '45%',
    width: 40,
    height: 40,
    borderRadius: 50,
    backgroundColor: COLORS.lunesFunctionalIncorrectDark,
    justifyContent: 'center',
    alignItems: 'center'
  },
  shadow: {
    shadowColor: COLORS.shadow,
    shadowOffset: { width: 0, height: 5 },
    shadowRadius: 3,
    shadowOpacity: 10
  }
})
const AnswerSection = ({
  currentDocumentNumber,
  setCurrentDocumentNumber,
  finishExercise,
  tryLater,
  trainingSet,
  disciplineTitle,
  documents
}: IAnswerSectionProps) => {
  const touchable: any = React.createRef()
  const [isPopoverVisible, setIsPopoverVisible] = useState(false)
  const [isActive, setIsActive] = useState(false)
  const [input, setInput] = useState('')
  const [submission, setSubmission] = useState('')
  const [result, setResult] = useState('')
  const [secondAttempt, setSecondAttempt] = useState(false)
  const document = documents[currentDocumentNumber]
  const totalNumbers = documents.length
  const [isFocused, setIsFocused] = useState(false)

  React.useEffect(() => {
    const _onSoundPlayerFinishPlaying = SoundPlayer.addEventListener('FinishedPlaying', () => setIsActive(false))

    const _onTtsFinishPlaying = Tts.addEventListener('tts-finish', () => setIsActive(false))

    return () => {
      _onSoundPlayerFinishPlaying.remove()
      _onTtsFinishPlaying.remove()
    }
  }, [])

  const checkEntry = () => {
    setSubmission(input)
    const splitInput = input.trim().split(' ')

    if (splitInput.length < 2) {
      setIsPopoverVisible(true)
      return
    }

    const article = splitInput[0].toLowerCase()
    const word = splitInput[1]

    if (!validateForSimilar(article, word)) {
      setResult('incorrect')
      storeResult('incorrect')
    } else if (validateForCorrect(article, word)) {
      setResult('correct')
      storeResult('correct')
    } else if (secondAttempt) {
      setResult('similar')
      storeResult('similar')
    } else {
      setInput('')
      setSecondAttempt(true)
      return
    }
    setSecondAttempt(false)
  }

  const validateForCorrect = (inputArticle: string, inputWord: string) => {
    const exactAnswer = inputArticle === document?.article && inputWord === document?.word

    const altAnswer = document?.alternatives?.some(
      // eslint-disable-next-line @typescript-eslint/naming-convention
      ({ article, alt_word }) => inputArticle === article && inputWord === alt_word
    )
    return exactAnswer || altAnswer
  }

  const validateForSimilar = (inputArticle: string, inputWord: string) => {
    const origCheck =
      document &&
      inputArticle === document.article &&
      stringSimilarity.compareTwoStrings(inputWord, document.word) > 0.4

    const altCheck = document?.alternatives?.some(
      // eslint-disable-next-line @typescript-eslint/naming-convention
      ({ article, alt_word }) =>
        inputArticle === article && stringSimilarity.compareTwoStrings(inputWord, alt_word) > 0.4
    )
    return origCheck || altCheck
  }

  const getNextWord = () => {
    setResult('')
    setInput('')
    setSecondAttempt(false)

    if (currentDocumentNumber === totalNumbers - 1) {
      finishExercise()
    }
    setCurrentDocumentNumber(currentDocumentNumber + 1)
  }

  const storeResult = async (score: string) => {
    await AsyncStorage.getItem('Vocabulary Trainer').then(value => {
      const jsonValue = value ? JSON.parse(value) : {}

      if (!jsonValue?.[disciplineTitle]?.[trainingSet]) {
        if (!jsonValue?.[disciplineTitle]) {
          jsonValue[disciplineTitle] = {}
        }
        jsonValue[disciplineTitle][trainingSet] = {}
      }

      if (document) {
        jsonValue[disciplineTitle][trainingSet][document.word] = {
          ...document,
          result: score
        }
      }

      AsyncStorage.setItem('Vocabulary Trainer', JSON.stringify(jsonValue))

      AsyncStorage.getItem('session').then(session => {
        const jsValue = JSON.parse(session)
        const newData = JSON.stringify({
          ...jsValue,
          retryData: {
            data: documents.slice(currentDocumentNumber + 1, totalNumbers)
          }
        })
        AsyncStorage.setItem('session', newData)
      })
    })
  }

  const handleSpeakerClick = (audio?: string) => {
    setIsActive(true)

    // Don't use soundplayer for IOS, since IOS doesn't support .ogg files
    if (audio && Platform.OS !== 'ios') {
      // audio from API
      SoundPlayer.playUrl(document?.audio)
    } else {
      Tts.speak(`${document?.article} ${document?.word}`, {
        androidParams: {
          KEY_PARAM_PAN: -1,
          KEY_PARAM_VOLUME: 0.5,
          KEY_PARAM_STREAM: 'STREAM_MUSIC'
        }
      })
    }
  }

  const getBorderColor = () => {
    if (isFocused) {
      return COLORS.lunesBlack
    } else if (!secondAttempt && !input) {
      return COLORS.lunesGreyMedium
    } else if (!result && !secondAttempt) {
      return COLORS.lunesBlack
    } else if (result === 'correct') {
      return COLORS.lunesFunctionalCorrectDark
    } else if (result === 'incorrect' || !secondAttempt) {
      return COLORS.lunesFunctionalIncorrectDark
    } else {
      return COLORS.lunesFunctionalAlmostCorrectDark
    }
  }

  const volumeIconColor =
    result === '' && !secondAttempt ? COLORS.lunesBlackUltralight : isActive ? COLORS.lunesRed : COLORS.lunesRedDark

  const volumeIconStyle = [styles.volumeIcon, (result !== '' || secondAttempt) && !isActive && styles.shadow]

  return (
    <View style={styles.container}>
      <Popover isVisible={isPopoverVisible} setIsPopoverVisible={setIsPopoverVisible} ref={touchable}>
        <PopoverContent />
      </Popover>

      <TouchableOpacity
        testID='volume-button'
        disabled={result === '' && !secondAttempt}
        style={volumeIconStyle}
        onPress={() => handleSpeakerClick(document?.audio)}>
        <VolumeUp fill={volumeIconColor} />
      </TouchableOpacity>

      <View
        testID='input-field'
        ref={touchable}
        style={[
          styles.textInputContainer,
          {
            borderColor: getBorderColor()
          }
        ]}>
        <TextInput
          style={styles.textInput}
          placeholder={secondAttempt ? 'Try again' : 'Enter Word with article'}
          placeholderTextColor={COLORS.lunesBlackLight}
          value={input}
          onChangeText={text => setInput(text)}
          editable={result === ''}
          onFocus={() => setIsFocused(true)}
          onBlur={() => setIsFocused(false)}
        />
        {(isFocused || (result === '' && input !== '')) && (
          <TouchableOpacity onPress={() => setInput('')}>
            <CloseIcon />
          </TouchableOpacity>
        )}
      </View>

      <Feedback secondAttempt={secondAttempt} result={result} document={document} input={submission} />

      <Actions
        tryLater={tryLater}
        giveUp={async () => {
          await storeResult('incorrect')
          getNextWord()
        }}
        input={input}
        result={result}
        checkEntry={checkEntry}
        getNextWord={getNextWord}
        secondAttempt={secondAttempt}
        isFinished={currentDocumentNumber === totalNumbers - 1}
      />
    </View>
  )
}

export default AnswerSection
