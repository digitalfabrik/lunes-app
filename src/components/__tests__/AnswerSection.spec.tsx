import React from 'react'
import AnswerSection from '../AnswerSection'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { IAnswerSectionProps } from '../../interfaces/exercise'

describe('Components', () => {
  describe('AnswerSection', () => {
    const defaultAnswerSectionProps: IAnswerSectionProps = {
      documents: [],
      profession: '',
      subCategory: '',
      currentDocumentNumber: 0,
      tryLater: () => {},
      finishExercise: () => {},
      setCurrentDocumentNumber: () => {}
    }

    it('renders correctly across screens', () => {
      const component = shallow(<AnswerSection {...defaultAnswerSectionProps} />)
      expect(toJson(component)).toMatchSnapshot()
    })

    it('should render volume button', () => {
      const component = shallow(<AnswerSection {...defaultAnswerSectionProps} />)
      expect(component.find('[testID="volume-button"]')).toHaveLength(1)
    })

    it('should render text input', () => {
      const component = shallow(<AnswerSection {...defaultAnswerSectionProps} />)
      expect(component.find('[testID="input-field"]')).toHaveLength(1)
    })
  })
})
