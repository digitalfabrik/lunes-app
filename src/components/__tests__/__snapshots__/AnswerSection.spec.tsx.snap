// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components AnswerSection renders correctly across screens 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": 867,
      "paddingVertical": 40,
      "position": "relative",
      "width": "100%",
    }
  }
>
  <ForwardRef
    isVisible={false}
    setIsPopoverVisible={[Function]}
  >
    <PopoverContent />
  </ForwardRef>
  <ForwardRef
    disabled={true}
    onPress={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#ff5252",
          "borderRadius": 50,
          "height": 40,
          "justifyContent": "center",
          "left": "45%",
          "position": "absolute",
          "top": -20,
          "width": 40,
        },
        false,
      ]
    }
    testID="volume-button"
  >
    <[object Object]
      fill="#e0e4ed"
    />
  </ForwardRef>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 2,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 106.5,
          "justifyContent": "space-between",
          "marginBottom": 80,
          "paddingHorizontal": 15,
          "width": 600,
        },
        Object {
          "borderColor": "#5b616a",
        },
      ]
    }
    testID="input-field"
  >
    <Component
      allowFontScaling={true}
      editable={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="Enter Word with article"
      placeholderTextColor="#777e91"
      rejectResponderTermination={true}
      style={
        Object {
          "color": "#000e38",
          "fontFamily": "SourceSansPro-Regular",
          "fontSize": 34,
          "fontWeight": "normal",
          "letterSpacing": 0.11,
          "width": 450,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <Feedback
    input=""
    result=""
    secondAttempt={false}
  />
  <Actions
    checkEntry={[Function]}
    getNextWord={[Function]}
    giveUp={[Function]}
    input=""
    isFinished={false}
    result=""
    secondAttempt={false}
    tryLater={[Function]}
  />
</View>
`;
